lib/steep/type_construction.rb

```diff
@@ -1644,8 +1644,7 @@ def synthesize(node, hint: nil, condition: false)
               if hint && !(tuples = select_flatten_types(hint) {|type| type.is_a?(AST::Types::Tuple) }).empty?
                 tuples.each do |tuple|
                   typing.new_child(node_range) do |child_typing|
-                    pair = with_new_typing(child_typing).try_tuple_type(node, tuple)
-                    if pair && pair.constr.check_relation(sub_type: pair.type, super_type: hint).success?
```

Comment:
I don't remember but it seems like the `check_relation` is to find the best tuple types from union types.

```ruby
# @type var tuple: [Integer, String] | [1, "foo"]
tuple = [1, ""]
```

The implementation doesn't work as expected because it just returns the first matching type, not the best one. The first-check-first-return strategy doesn't make any sense because the order in union types is not preserved.